File: golang/length_of_lis.go
package main

import "fmt"

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
func lengthOfLIS(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	s := make([]int, len(nums))
	for i := range s {
		s[i] = 1
	}
	var i, j int
	for i = 1; i < len(nums); i++ {
		for j = i; j >= 0; j-- {
			if nums[i] > nums[j] {
				s[i] = max(s[j]+1, s[i])
			}
		}
	}
	fmt.Println(s)
	var result int
	for i := 0; i < len(s); i++ {
		if result < s[i] {
			result = s[i]
		}
	}
	return result
}

func main() {
	a := []int{10, 9, 2, 5, 3, 7, 101, 18}
	fmt.Println(lengthOfLIS(a))
}

File: rust/length_of_lis.rs
use std::cmp::max;
pub fn length_of_lis(nums: Vec<i32>) -> i32 {
    let mut dp = vec![1; nums.len()];

    for i in 0..nums.len() {
        for j in 0..i {
            if nums[i] > nums[j] {
                dp[i] = max(dp[i], dp[j] + 1)
            }
        }
    }

    match dp.iter().max() {
        None => 0,
        Some(&x) => x,
    }
}
fn main() {
    let v = vec![100, 32, 57, 17, 75, 45, 108];
    println!("{}", length_of_lis(v))
}
